In [ ]:
#import
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
train = pd.read_csv('train.csv')

train.head(3)

ID  나이  키(cm)  몸무게(kg)    BMI    시력  충치  공복 혈당  혈압  중성 지방  \
0  TRAIN_0000  35    170       70  24.22  1.10   1     98  40     80   
1  TRAIN_0001  40    150       55  24.44  1.00   0    173  39    104   
2  TRAIN_0002  60    170       50  17.30  0.75   0     96  40     61   

   혈청 크레아티닌  콜레스테롤  고밀도지단백  저밀도지단백  헤모글로빈  요 단백  간 효소율  label  
0       1.3    211      75     120   15.9     1   1.53      1  
1       0.6    251      46     184   11.8     1   1.45      0  
2       0.8    144      43      89   15.3     1   1.04      0

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7000 entries, 0 to 6999
Data columns (total 18 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        7000 non-null   object 
 1   나이        7000 non-null   int64  
 2   키(cm)     7000 non-null   int64  
 3   몸무게(kg)   7000 non-null   int64  
 4   BMI       7000 non-null   float64
 5   시력        7000 non-null   float64
 6   충치        7000 non-null   int64  
 7   공복 혈당     7000 non-null   int64  
 8   혈압        7000 non-null   int64  
 9   중성 지방     7000 non-null   int64  
 10  혈청 크레아티닌  7000 non-null   float64
 11  콜레스테롤     7000 non-null   int64  
 12  고밀도지단백    7000 non-null   int64  
 13  저밀도지단백    7000 non-null   int64  
 14  헤모글로빈     7000 non-null   float64
 15  요 단백      7000 non-null   int64  
 16  간 효소율     7000 non-null   float64
 17  label     7000 non-null   int64  
dtypes: float64(5), int64(12), object(1)
memory usage: 984.5+ KB


In [ ]:
train.describe()



나이        키(cm)      몸무게(kg)          BMI           시력  \
count  7000.000000  7000.000000  7000.000000  7000.000000  7000.000000   
mean     43.973571   164.781429    65.932857    24.144423     1.011650   
std      12.063793     9.170213    12.978702     3.501945     0.427828   
min      20.000000   135.000000    30.000000    14.270000     0.100000   
25%      35.000000   160.000000    55.000000    21.600000     0.800000   
50%      40.000000   165.000000    65.000000    23.880000     1.000000   
75%      50.000000   170.000000    75.000000    26.120000     1.200000   
max      85.000000   190.000000   130.000000    42.450000     9.900000   

                충치        공복 혈당           혈압        중성 지방     혈청 크레아티닌  \
count  7000.000000  7000.000000  7000.000000  7000.000000  7000.000000   
mean      0.227429    99.331857    45.532857   127.144714     0.884900   
std       0.419202    21.129670     8.820611    73.918492     0.241523   
min       0.000000    57.000000    14.000000    21.000000     0.100000   
25%       0.000000    89.000000    40.000000    74.000000     0.800000   
50%       0.000000    96.000000    45.000000   107.000000     0.900000   
75%       0.000000   104.000000    50.000000   161.000000     1.000000   
max       1.000000   386.000000    91.000000   999.000000    10.000000   

             콜레스테롤       고밀도지단백       저밀도지단백        헤모글로빈         요 단백  \
count  7000.000000  7000.000000  7000.000000  7000.000000  7000.000000   
mean    197.276571    57.355429   115.346857    14.631914     1.083857   
std      36.306494    14.506945    41.788153     1.540907     0.392051   
min      86.000000    18.000000     1.000000     4.900000     1.000000   
25%     173.000000    47.000000    92.000000    13.600000     1.000000   
50%     195.000000    55.000000   113.000000    14.800000     1.000000   
75%     219.000000    66.000000   136.000000    15.700000     1.000000   
max     395.000000   157.000000  1340.000000    20.900000     5.000000   

             간 효소율        label  
count  7000.000000  7000.000000  
mean      1.144696     0.367286  
std       0.432735     0.482100  
min       0.140000     0.000000  
25%       0.840000     0.000000  
50%       1.100000     0.000000  
75%       1.380000     1.000000  
max       5.670000     1.000000

#키와 몸무게는 흡연과 큰 관계가 없을 것이다. 그렇기 때문에 이 두가지 정보를 포함한 BMI는 삭제하는 것이 적절할 것 같다.

In [ ]:
# 데이터셋에서 BMI 열 제거
train.drop(columns=['BMI'], inplace=True)


NameError: name 'train' is not defined

# 변수들에 대한 이상치 검사

시력 : 최대값이 9.9이므르 이상치처리 필요

공복 혈당 : 정상 수치가 70~100임, 126이상일 시 당뇨병으로 분류 - 담배와 관련 있음 -> 이상치 처리에 신중해야 함 : 이상치 처리를 안해도 될 거 같다.

혈압 : 혈압의 경우 이완기 수축기가 있는데 해당 데이터셋은 이완기 혈압으로 추측됨, 정상의 경우 80, 저혈압은 60이하, 고혈압은 90이상이라 함 - 담배와 관련 있음

중성지방 : 150미만 정상, 150 ~ 199 경계, 200 ~ 499 높음, 500 ~ 아주 높음 - 999가 있을 수 있는 수치만 75%와 비교했을 때 너무 높기 때문에 스케일링이 필요해 보임 - 담배와 관련 있음

혈청 크레아티닌 : 정상 범위 남성 0.7~1.4, 여성 0.5 ~ 1.1

콜레스테롤 : 200미만 정상, 200~239 경계, 240 이상 고콜레스테롤 혈증 - 이상치 처리가 필요 없어보임


고밀도지단백 : 60이상이면 정상 높으면 높을수로 좋다고 함.

저밀도지단백 : 130미만 정상, 130 ~ 159 경계, 160 이상 고콜레스테롤 혈증

헤모글로빈 : 남성 13.5-17.5 정상, 여성 12.5-15.5 정상 - 이상치 처리 필요 없어보임

요단백 - 이상처 처리 필요없어보임

간 효소율

시력의 이상치 검사

- 시력의 경우 흡연 여부와 큰 연관이 없을 것이라 판단해 이상치 처리를 중간값으로 할 예정

In [ ]:
#시력의 이상치 검사 - IQR(사분위수범위) : 적절하지 않은 것 같다
Q1 = train['시력'].quantile(0.25)
Q3 = train['시력'].quantile(0.75)
IQR = Q3 - Q1
outliers = train['시력'][(train['시력'] < (Q1 - 1.5 * IQR)) | (train['시력'] > (Q3 + 1.5 * IQR))]
print(outliers)

42      2.00
138     2.00
148     0.20
355     5.70
403     2.00
        ... 
6634    0.20
6639    2.00
6662    0.20
6733    0.20
6788    5.45
Name: 시력, Length: 95, dtype: float64


In [ ]:
# 시력의 이상치 검사 - z-score : 적절한 것 같다

시력_평균 = train['시력'].mean()
시력_표준편차 = train['시력'].std()
시력_중간값 = train['시력'].median()
threshold = 3

outlier = []

for i in train['시력']:
    z = (i-시력_평균)/시력_표준편차
    if z > threshold:
        outlier.append(i)

outlier

In [ ]:
#시력 이상치 처리 - 중간값으로

train['시력'] = train['시력'].apply(lambda x: 시력_중간값 if (x - 시력_평균) / 시력_표준편차 > threshold else x)



공복 혈당의 이상치 검사

In [ ]:
#공복 혈당의 이상치 검사 - IQR(사분위수범위) :
Q1 = train['공복 혈당'].quantile(0.25)
Q3 = train['공복 혈당'].quantile(0.75)
IQR = Q3 - Q1
outliers = train['공복 혈당'][(train['공복 혈당'] < (Q1 - 1.5 * IQR)) | (train['공복 혈당'] > (Q3 + 1.5 * IQR))]
print(outliers)

1       173
39      187
40      129
110     129
196     137
       ... 
6949    154
6953    127
6975    233
6982    308
6986    268
Name: 공복 혈당, Length: 402, dtype: int64


In [ ]:
#공복 혈당의 이상치 검사 - z-score

공복_혈당_평균 = train['공복 혈당'].mean()
공복_혈당_표준편차 = train['공복 혈당'].std()
공복_혈당_중간값 = train['공복 혈당'].median()
threshold = 3

outlier = []

for i in train['공복 혈당']:
    z = (i-공복_혈당_평균)/공복_혈당_표준편차
    if z > threshold:
        outlier.append(i)

outlier

[173,
 187,
 191,
 274,
 164,
 222,
 167,
 201,
 247,
 285,
 165,
 205,
 288,
 238,
 187,
 229,
 238,
 184,
 175,
 252,
 181,
 164,
 234,
 244,
 182,
 218,
 342,
 375,
 282,
 199,
 168,
 247,
 177,
 164,
 188,
 230,
 168,
 165,
 173,
 189,
 166,
 205,
 165,
 195,
 250,
 232,
 185,
 164,
 164,
 189,
 207,
 230,
 237,
 271,
 206,
 197,
 181,
 197,
 167,
 237,
 272,
 182,
 229,
 201,
 241,
 228,
 213,
 171,
 170,
 243,
 386,
 314,
 171,
 243,
 228,
 240,
 187,
 167,
 168,
 184,
 184,
 302,
 224,
 200,
 182,
 256,
 185,
 169,
 238,
 221,
 201,
 313,
 163,
 205,
 212,
 195,
 199,
 197,
 163,
 164,
 290,
 163,
 164,
 201,
 164,
 233,
 173,
 258,
 185,
 198,
 251,
 216,
 256,
 174,
 198,
 170,
 258,
 233,
 171,
 185,
 177,
 233,
 308,
 268]

In [ ]:
#혈압의 이상치 검사 - IQR(사분위수범위) : 적절하지 않은 것 같다
Q1 = train['혈압'].quantile(0.25)
Q3 = train['혈압'].quantile(0.75)
IQR = Q3 - Q1
outliers = train['혈압'][(train['혈압'] < (Q1 - 1.5 * IQR)) | (train['혈압'] > (Q3 + 1.5 * IQR))]
print(outliers)

17      70
69      73
76      72
176     71
232     72
        ..
6900    74
6926    23
6946    70
6958    70
6976    81
Name: 혈압, Length: 176, dtype: int64


In [ ]:
# 혈압의 이상치 검사 - z-score : 적절하지 않은 것 같음

혈압_평균 = train['혈압'].mean()
혈압_표준편차 = train['혈압'].std()
혈압_중간값 = train['혈압'].median()
threshold = 3

outlier = []

for i in train['혈압']:
    z = abs((i-혈압_평균)/혈압_표준편차 )
    if z > threshold:
        outlier.append(i)


outlier


[73,
 72,
 72,
 72,
 17,
 80,
 74,
 91,
 75,
 73,
 73,
 74,
 82,
 77,
 79,
 79,
 75,
 77,
 89,
 76,
 88,
 80,
 74,
 19,
 72,
 82,
 74,
 72,
 78,
 74,
 80,
 78,
 14,
 82,
 75,
 82,
 72,
 85,
 75,
 79,
 74,
 80,
 73,
 77,
 82,
 17,
 72,
 74,
 81]

혈압의 경우 극단적으로 높은 값보다 낮은 값의 이상치가 있어 LCL로 처리

In [ ]:
k = 3  # 일반적으로 3으로 설정

LCL = 혈압_평균 - k * 혈압_표준편차

print(f' LCL: {LCL}')


 LCL: 19.071024212291086


In [ ]:
# 혈압은 흡연과 관계가 유의미하다고 판단되어 LCL값으로 대체함
train['혈압'] = train['혈압'].apply(lambda x: LCL if ( x < LCL) else x)


In [ ]:
train['혈압'].min()

19.071024212291086

In [ ]:
#중성지방 사분위수 - 적절하지 않음

Q1 = train['중성 지방'].quantile(0.25)
Q3 = train['중성 지방'].quantile(0.75)
IQR = Q3 - Q1
outliers = train['중성 지방'][(train['중성 지방'] < (Q1 - 1.5 * IQR)) | (train['중성 지방'] > (Q3 + 1.5 * IQR))]
print(outliers.sort_values())


2117    292
6331    292
2703    292
353     292
3001    292
       ... 
5390    397
5947    397
1273    398
5796    399
5839    999
Name: 중성 지방, Length: 296, dtype: int64


In [ ]:
# 중성 지방 z - score - 적절하지 않은 것 같음

중성_지방_평균 = train['중성 지방'].mean()
중성_지방_표준편차 = train['중성 지방'].std()
중성_지방_중간값 = train['중성 지방'].median()


outlier = []

for i in train['중성 지방']:
    z = abs((i-중성_지방_평균)/중성_지방_표준편차 )
    if z > threshold:
        outlier.append(i)


outlier

[351,
 356,
 367,
 388,
 387,
 386,
 388,
 376,
 370,
 385,
 351,
 375,
 354,
 383,
 362,
 368,
 389,
 387,
 378,
 393,
 352,
 372,
 398,
 377,
 397,
 374,
 376,
 351,
 375,
 362,
 363,
 394,
 367,
 370,
 382,
 388,
 384,
 361,
 350,
 387,
 358,
 393,
 365,
 394,
 379,
 394,
 370,
 353,
 358,
 362,
 353,
 369,
 386,
 377,
 381,
 355,
 364,
 384,
 353,
 371,
 372,
 353,
 360,
 361,
 381,
 391,
 381,
 364,
 357,
 376,
 381,
 373,
 349,
 362,
 354,
 379,
 352,
 357,
 391,
 358,
 396,
 349,
 385,
 377,
 397,
 397,
 356,
 378,
 369,
 362,
 399,
 999,
 369,
 397,
 371,
 364,
 370,
 372,
 381,
 397,
 382,
 365,
 368,
 387,
 365,
 397,
 370,
 381,
 387,
 354,
 385,
 391]

In [ ]:
# LCL, UCL 방식 모두 적절하지 않은 것 같다.

k = 3  # 일반적으로 3으로 설정
UCL = 중성_지방_평균 + k * 중성_지방_표준편차
LCL = abs(중성_지방_평균 - k * 중성_지방_표준편차)

print(f'UCL: {UCL}, LCL: {LCL}')


UCL: 348.9001906655585, LCL: 94.61076209412997


In [ ]:
train['중성 지방'] = train['중성 지방'].apply(lambda x: 600 if ( x > 600) else x)

In [ ]:
#혈청 크레아티닌 사분위수 - 적절하지 않음

Q1 = train['혈청 크레아티닌'].quantile(0.25)
Q3 = train['혈청 크레아티닌'].quantile(0.75)
IQR = Q3 - Q1

outliers = train['혈청 크레아티닌'][(train['혈청 크레아티닌'] < (Q1 - 1.5 * IQR)) | (train['혈청 크레아티닌'] > (Q3 + 1.5 * IQR))]
print(outliers.sort_values())

4575     0.1
1035     0.1
2164     0.3
2248     0.4
4365     0.4
        ... 
5914     2.5
4492     2.6
3498     5.9
2335     7.4
3965    10.0
Name: 혈청 크레아티닌, Length: 395, dtype: float64


In [ ]:
# 혈청 크레아티닌 UCL

혈청_크레아티닌_평균 = train['혈청 크레아티닌'].mean()
혈청_크레아티닌_표준편차 = train['혈청 크레아티닌'].std()


k = 3  # 일반적으로 3으로 설정
UCL = 혈청_크레아티닌_평균 + k * 혈청_크레아티닌_표준편차
LCL = abs(혈청_크레아티닌_평균 - k * 혈청_크레아티닌_표준편차)

print(f'UCL: {UCL}, LCL: {LCL}')


UCL: 1.6094678878401634, LCL: 0.16033211215983645


In [ ]:
train['혈청 크레아티닌'] = train['혈청 크레아티닌'].apply(lambda x: UCL if ( x > UCL) else x)

In [ ]:
train['혈청 크레아티닌'].max()

1.6094678878401634

In [ ]:
#고밀도지단백의 이상치 검사

고밀도지단백_평균 = train['고밀도지단백'].mean()
고밀도지단백_표준편차 = train['고밀도지단백'].std()
k = 3
UCL = 고밀도지단백_평균 + k * 고밀도지단백_표준편차
LCL = abs(고밀도지단백_평균 - k * 고밀도지단백_표준편차)

print(f'UCL: {UCL}, LCL: {LCL}')

outlier = []


for i in train['고밀도지단백']:
    z = abs((i-고밀도지단백_평균)/고밀도지단백_표준편차 )
    if z > threshold:
        outlier.append(i)


outlier

UCL: 100.87626316569808, LCL: 13.834593977159066


[]

In [ ]:
#고밀도지단백 이상치 처리
train['고밀도지단백'] = train['고밀도지단백'].apply(lambda x: UCL if ( x > UCL) else x)

In [ ]:
#저밀도지단백의 이상치 검사

저밀도지단백_평균 = train['저밀도지단백'].mean()
저밀도지단백_표준편차 = train['저밀도지단백'].std()
k = 3
UCL = 저밀도지단백_평균 + k * 저밀도지단백_표준편차
LCL = abs(저밀도지단백_평균 - k * 저밀도지단백_표준편차)

print(f'UCL: {UCL}, LCL: {LCL}')

outlier = []


for i in train['저밀도지단백']:
    z = abs((i-저밀도지단백_평균)/저밀도지단백_표준편차 )
    if z > threshold:
        outlier.append(i)


outlier

UCL: 240.71131595067902, LCL: 10.01760166496473


[257, 1070, 307, 292, 251, 256, 242, 282, 1340, 245, 295, 590, 293, 1120, 910]

In [ ]:
#저밀도지단백 이상치 처리
train['저밀도지단백'] = train['저밀도지단백'].apply(lambda x: UCL if ( x > UCL) else x)

In [ ]:
train.describe()


나이        키(cm)      몸무게(kg)           시력           충치  \
count  7000.000000  7000.000000  7000.000000  7000.000000  7000.000000   
mean     43.973571   164.781429    65.932857     0.991986     0.227429   
std      12.063793     9.170213    12.978702     0.299851     0.419202   
min      20.000000   135.000000    30.000000     0.100000     0.000000   
25%      35.000000   160.000000    55.000000     0.800000     0.000000   
50%      40.000000   165.000000    65.000000     1.000000     0.000000   
75%      50.000000   170.000000    75.000000     1.200000     0.000000   
max      85.000000   190.000000   130.000000     2.000000     1.000000   

             공복 혈당           혈압        중성 지방     혈청 크레아티닌        콜레스테롤  \
count  7000.000000  7000.000000  7000.000000  7000.000000  7000.000000   
mean     99.331857    45.534183   127.087714     0.881758   197.276571   
std      21.129670     8.816353    73.398098     0.189621    36.306494   
min      57.000000    19.071024    21.000000     0.100000    86.000000   
25%      89.000000    40.000000    74.000000     0.800000   173.000000   
50%      96.000000    45.000000   107.000000     0.900000   195.000000   
75%     104.000000    50.000000   161.000000     1.000000   219.000000   
max     386.000000    91.000000   600.000000     1.609468   395.000000   

            고밀도지단백       저밀도지단백        헤모글로빈         요 단백        간 효소율  \
count  7000.000000  7000.000000  7000.000000  7000.000000  7000.000000   
mean     57.265047   114.755524    14.631914     1.083857     1.144696   
std      14.163916    33.654860     1.540907     0.392051     0.432735   
min      18.000000     1.000000     4.900000     1.000000     0.140000   
25%      47.000000    92.000000    13.600000     1.000000     0.840000   
50%      55.000000   113.000000    14.800000     1.000000     1.100000   
75%      66.000000   136.000000    15.700000     1.000000     1.380000   
max     100.876263   240.711316    20.900000     5.000000     5.670000   

             label  
count  7000.000000  
mean      0.367286  
std       0.482100  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000

In [ ]:
train.to_csv('outlier_com_train.csv')

#이상치 처리 완료

###-나이 이상치 처리 x
###-키 이상치 처리 x
###-몸무게 이상치 처리 x
###-시력 이상치 처리 o - Z-score를 이용해 중간값으로 대체함
###-충치 이상치 처리 x
###-공복 혈당 이상치 처리 x
###-혈압 이상치 처리 o - LCL을 이용해 LCL값으로 대체함
###-중성 지방 이상치 처리 o - 600보다 큰값들을 600으로 대체함
###-혈청 크레아티닌 이상치 처리 o - UCL을 이용해 탐지 후 UCL값으로 대체함
###-콜레스테롤 이상치 처리 x
###-고밀도지단백 이상치 처리 o - UCL을 이용해 탐지 후 UCL값으로 대체함
###-저밀도지단백 이상치 처리 o - UCL을 이용해 탐지 후 UCL값으로 대체함
###-헤모글로빈 이상치 처리 x
###-요 단백	이상치 처리 x
###-간 효소율 이상치 처리 x

###BMI 삭제함
###콜레스테롤과 저밀도지단백과의 다중공선성 고려해야할듯